README

mbootpack
---------

This is a utility to take a multiboot kernel and modules and repackage
them in a form that a standard linux bootloader will be able to load them.
It statically allocates memory addresses based on a 'standard' PC memory 
layout, and then saves the image of the loaded system, along with an 
almost-standard linux bzImage header which takes care of the start-of-day 
requirements of a multiboot kernel (setting up 32-bit protected mode, etc.)

Example invocation, to package a xen VMM and xenlinux guest and initrd:

   mbootpack -o bzImage -m ./xenlinux -m ./initrd.img ./xen-image

You can now boot the 'bzImage' file using your favourite linux bootloader.

The kernel command line will be provided at boot time by the bootloader
(you can specify a kernel command-line using the '-c' flag, but it will
be overridden at boot time unless the bootloder provides an entirely
empty command line).  If you want to override the command line for the
first module (i.e. domain 0 kernel in Xen) at boot time, append ' -- '
and the module commadn line to the bootloader command line, e.g.:

  boot: bzImage com1=9600,8n1 console=com1 dom0_mem=49152 -- root=/dev/sda3 ro console=ttyS0,9600n8 

Everything before the '--' is passed to the kernel (xen) as its command
line; everything after is passed to the first module (xenlinux).

This is ALPHA code: there are execution paths which have *not* been
tested, though it works for loading the Xen hypervisor using GrUB, LILO
or SYSLINUX.  Bug reports and patches are very welcome.

Possible features for future versions (all look possible, if there's any 
demand for them):

    - support for kernels that load below 1MB
    - zImage-style compressed images
    - support for the MULTIBOOT_VIDEO_MODE bit
    - proper support for passing E820h memory-maps from bzImage


Tim Deegan <tjd@phlegethon.org>, April 2005



License and attributions
------------------------

The bzImage header block was originally taken from the Linux kernel.
http://www.kernel.org/

Some parts of the Multiboot loader code are based on GNU GRUB.
mb_info.h and mb_header.h are taken from GNU GRUB.
http://www.gnu.org/software/grub/

All other code is copyright (C) 2003-2017 Tim Deegan (tjd@phlegethon.org)

mbootpack is distributed under the GNU General Public License: see "GPL"

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


Changes
Version 0.7a (May 2017)
-----------------------

Changed loader to skip program headers that aren't PT_LOAD.

Fixed a few compile issues that cropped up in the last nine years.


Version 0.6a (December 2008)
----------------------------

Added '-t' flag to strip out paths when making module command-lines; it
should be specified before the '-m' arguments.

Fixed the layout for kernels that load above 1MB; they are now loaded at
1MB and relocated at boot time instead of having a lot of zeroes in the
output file.

Made 64-bit compiles work properly.

Ditched bin2c in favour of some "od" runes taken from the xen BIOS build.


Version 0.5a (January 2007)
---------------------------

Fixed a type-size error that made 64-bit builds choke on the ELF program
headers.


Version 0.4a (April 2005)
-------------------------

Added this change log. 

Added check to catch kernels loading at very high addresses.

Tidied makefile dependency generation so 'make clean' doesn't rebuild 
the setup sectors.


Version 0.3a (March 2005)
-------------------------

Added checks to detect and warn if the input files are compressed.

Changed kernel loader to ignore zero-length sections in ELF kernel phdrs
(GNU GRUB has one).


Version 0.2a (March 2005)
-------------------------

Changed the location of the MBI and associated data from just after the 
kernel to just below the 640k memory hole, because Xen assumes they will be 
below 1MB and might overwrite them otherwise.

Added support for providing the command-line at boot time, and for 
splitting the boot command-line between the kernel and the first module.

Removed pack-as-multiboot option, which was really a debugging feature.


Version 0.1a (November 2004)
----------------------------

Initial alpha release


bootsect.S
/*
 *  bootsect.S
 *
 *  This is bootsect.S from the linux 2.6.9 sources,
 *  with minor changes for mbootpack.
 *
 *  
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation; either version 2 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 *  02111-1307, USA.
 */
	
#include "mbootpack.h"
		
/*
 *	bootsect.S		Copyright (C) 1991, 1992 Linus Torvalds
 *
 *	modified by Drew Eckhardt
 *	modified by Bruce Evans (bde)
 *	modified by Chris Noe (May 1999) (as86 -> gas)
 *	gutted by H. Peter Anvin (Jan 2003)
 *
 * BIG FAT NOTE: We're in real mode using 64k segments.  Therefore segment
 * addresses must be multiplied by 16 to obtain their respective linear
 * addresses. To avoid confusion, linear addresses are written using leading
 * hex while segment addresses are written as segment:offset.
 *
 */
 
/* #include <asm/boot.h> */
/* Definitions we should have got from there */	
#define DEF_INITSEG     0x9000
#define DEF_SYSSEG      0x1000
#define DEF_SETUPSEG    0x9020
#define DEF_SYSSIZE     0x7F00
#define NORMAL_VGA      0xffff
#define EXTENDED_VGA    0xfffe
#define ASK_VGA         0xfffd


/* SETUPSECTS	= 4 */			/* default nr of setup-sectors */
BOOTSEG		= 0x07C0		/* original address of boot-sector */
INITSEG		= DEF_INITSEG		/* we move boot here - out of the way */
SETUPSEG	= DEF_SETUPSEG		/* setup starts here */
SYSSEG		= DEF_SYSSEG		/* system loaded at 0x10000 (65536) */
SYSSIZE		= DEF_SYSSIZE		/* system size: # of 16-byte clicks */
					/* to be loaded */
ROOT_DEV	= 0 			/* ROOT_DEV is now written by "build" */
SWAP_DEV	= 0			/* SWAP_DEV is now written by "build" */

#ifndef SVGA_MODE
/* #define SVGA_MODE ASK_VGA */
#define SVGA_MODE NORMAL_VGA
#endif

#ifndef RAMDISK
#define RAMDISK 0
#endif

#ifndef ROOT_RDONLY
#define ROOT_RDONLY 1
#endif

.code16
.text

.global _start
_start:

	# Normalize the start address
	jmpl	$BOOTSEG, $start2

start2:
	movw	%cs, %ax
	movw	%ax, %ds
	movw	%ax, %es
	movw	%ax, %ss
	movw	$0x7c00, %sp
	sti
	cld

	movw	$bugger_off_msg, %si

msg_loop:
	lodsb
	andb	%al, %al
	jz	die
	movb	$0xe, %ah
	movw	$7, %bx
	int	$0x10
	jmp	msg_loop

die:
	# Allow the user to press a key, then reboot
	xorw	%ax, %ax
	int	$0x16
	int	$0x19

	# int 0x19 should never return.  In case it does anyway,
	# invoke the BIOS reset code...
	ljmp	$0xf000,$0xfff0


bugger_off_msg:
	.ascii	"Direct booting from floppy is no longer supported.\r\n"
	.ascii	"Please use a boot loader program instead.\r\n"
	.ascii	"\n"
	.ascii	"Remove disk and press any key to reboot . . .\r\n"
	.byte	0
	

	# Kernel attributes; used by setupbegtext

	.org 497
setup_sects:	.byte SETUPSECTS
root_flags:	.word ROOT_RDONLY
syssize:	.word SYSSIZE
swap_dev:	.word SWAP_DEV
ram_size:	.word RAMDISK
vid_mode:	.word SVGA_MODE
root_dev:	.word ROOT_DEV
boot_flag:	.word 0xAA55


mbootpack.h
/*
 *  mbootpack.h
 *
 *  Common definitions for mbootpack
 * 
 *  Copyright (C) 2003-2017 Tim Deegan (tjd@phlegethon.org)
 * 
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation; either version 2 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 *  02111-1307, USA.
 */

#ifndef __MBOOTPACK__H__
#define __MBOOTPACK__H__

#ifndef __MB_ASM

#undef NDEBUG
#include <stdio.h>
#include <stdint.h>

/* Flags */
extern int quiet;

/* Types */
typedef uintptr_t address_t;

typedef struct section_t {
    char *buffer;
    address_t start;
    int32_t size;
    struct section_t *prev;
    struct section_t *next;
} section_t;

/* buildimage.c */
extern void make_bzImage(section_t *sections, 
                         address_t entry, 
                         address_t mbi, 
                         FILE *fp);

address_t place_mbi(long int size);


/* trampoline.S */
extern unsigned char mb_trampoline[];
extern unsigned char mb_trampoline_end[];
extern volatile address_t mb_mbi_address, mb_entry_address;

/* Macros */
#define MIN(_x,_y) (((_x)<=(_y))?(_x):(_y))
#define MAX(_x,_y) (((_x)<=(_y))?(_y):(_x))
#define ROUNDUP_P2(_x, _a) (((_x)+((_a)-1))&(~((_a)-1)))

#endif

/* x86 memory: such fun */
#define MEM_HOLE_START  0xa0000
#define MEM_HOLE_END    0x100000
#define HIGHMEM_START   MEM_HOLE_END
#define X86_PAGE_SIZE   0x1000

/* How much command line we'll take from the bootloader. */
#define CMD_LINE_SPACE  0x300

/* Number of 512-byte sectors to load in low memory (max 7) */
#define SETUPSECTS	7


/* Who are we? */
#define MBOOTPACK_VERSION_STRING "v0.7 (alpha)"

#endif /* __MBOOTPACK__H__ */

/*
 *  EOF (mbootpack.h)
 */



trampoline.S
/*
 *  trampoline.S
 *
 *  Brief setup code for the multiboot output format
 * 
 *  Copyright (C) 2004  Tim Deegan (tjd21@cl.cam.ac.uk)
 * 
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation; either version 2 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 *  02111-1307, USA.
 *
 * $Id: trampoline.S,v 1.2 2004/11/10 09:53:42 tjd21 Exp $
 *
 */

.globl mb_trampoline, mb_trampoline_end, mb_mbi_address, mb_entry_address

.code32
.data   /* Not .text, because we need to be able to write to it */
		
/* This code is called by the *real* multiboot bootloader.
 * It copies some values into the MBI struct and jumps at the
 * kernel.  Twice here there's a load-immediate instruction
 * fabricated from bytes so we can alter it at link time.  Ugh. */
	
mb_trampoline:	
	/* Copy some things from the *old* multiboot struct */
	movl	%ebx, %ecx
	/* ... to the *new* multiboot information struct */
	.byte	0xbb
mb_mbi_address:	
	.long	0xdeadbeef	#  == mov mbi_address, %ebx
	movl	0(%ecx), %eax
	andl	$0x1, %eax
	jz	no_memfields
	orl	$0x1, 0(%ebx) 
	movl	4(%ecx), %eax
	movl	%eax, 4(%ebx)
	movl	8(%ecx), %eax
	movl	%eax, 8(%ebx)
no_memfields:
	
	movl	0(%ecx), %eax
	andl	$0x2, %eax
	jz	no_bootdev
	orl	$0x2, 0(%ebx) 
	movl	12(%ecx), %eax
	movl	%eax, 12(%ebx)
no_bootdev:

	movl	0(%ecx), %eax
	andl	$0x40, %eax
	jz	no_memmap
	orl	$0x40, 0(%ebx)
	movl	44(%ecx), %eax
	movl	%eax, 44(%ebx)
	movl	48(%ecx), %eax
	movl	%eax, 48(%ebx)
no_memmap:

	/* Magic multiboot number */
	movl	$0x2BADB002, %eax
	/* Launch the kernel */
	.byte	0xb9
mb_entry_address:	
	.long	0xdeadbeef	# == mov entry_address, %ecx
	jmp	*%ecx
mb_trampoline_end:

foo:	nop
	jmp foo
	
/*
 *  EOF (trampoline.S)
 */


buildimage.c
/*
 *  buildimage.c
 *
 *  Takes the memory image of a loaded kernel and modules and repackages 
 *  it as a linux bzImage
 *
 *  Copyright (C) 2003-2017 Tim Deegan (tjd@phlegethon.org)
 * 
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License as
 *  published by the Free Software Foundation; either version 2 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 *  02111-1307, USA.
 */



#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <string.h>
#include <getopt.h>
#include <elf.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/mman.h>

#include "mbootpack.h"
#include "mb_header.h"

/*  We will build an image that a bzImage-capable bootloader will load like 
 *  this:
 * 
 *  ==============   (0)
 *  (BIOS memory)
 *  --------------
 *  (Bootloader)
 *  --------------
 *  bzImage startup code
 *  MBI, command-lines, module info
 *  ==============   (0xa0000)
 *  (memory hole)
 *  ==============   (0x100000)
 *  Kernel and modules
 *  ==============
 * 
 *  The bzImage startup code is mostly taken straight from the linux kernel
 *  (see bootsect.S, startup.S).  It does the usual unpleasant start-of-day
 *  tasks to get to 32-bit protected mode, then sets registers appropriately 
 *  and jumps to the kernel's entry address.
 *  
 *  It also does some relocation to make sure the MBI is where we expect it, 
 *  and parses the linux command line.
 */

#define BZ_SETUP_OFFSET       (512 * (1 + SETUPSECTS)) 
/* These *MUST* fit the offsets of entry_address and mbi_address in setup.S */
#define BZ_ENTRY_OFFSET       0x30
#define BZ_MBI_OFFSET         0x34
#define BZ_HIGH_START_OFFSET  0x38
#define BZ_HIGH_SIZE_OFFSET   0x3C

/* Bring in the bzImage boot sector and setup code */
#include "bzimage_header.c"

address_t place_mbi(long int size) 
/* Find space at the top of *low* memory for the MBI and associated red tape */
{
    address_t start;
    start = 0xa000 - size;
    if (start < 0x9000 + sizeof(bzimage_bootsect) + sizeof(bzimage_setup)) {
        printf("Fatal: command-lines too long: need %i, have %i bytes\n",
               size, 
               0x1000 - (sizeof(bzimage_bootsect) + sizeof(bzimage_setup)));
        exit(1);        
    }
    if (!quiet) {
        printf("Placed MBI and strings (%p+%p)\n", 
               start, size);
    }
    return start;
}

void make_bzImage(section_t *sections, 
                  address_t entry, 
                  address_t mbi,
                  FILE *fp)
/* Rework this list of sections into a bzImage and write it out to fp */
{
    int i;
    size_t offset;
    section_t *s;
    address_t highmem_start = 0xffffffff, highmem_size = 0;

    /* Patch the kernel and mbi addresses into the setup code */
    *(address_t *)(bzimage_setup + BZ_ENTRY_OFFSET) = entry;
    *(address_t *)(bzimage_setup + BZ_MBI_OFFSET) = mbi;
    if (!quiet) printf("Kernel entry is %p, MBI is %p.\n", entry, mbi);

    /* If the high-memory sections don't start at 1MB, shuffle them down 
     * in the file and record how they should be relocated at boot time. */
    for (s = sections, i = 0; s; s = s->next) {
	if (s->start < HIGHMEM_START) continue;
	highmem_start = MIN(highmem_start, s->start);
	highmem_size = MAX(highmem_size, s->start + s->size - highmem_start);
    }
    if (highmem_size != 0) {
	*(address_t *)(bzimage_setup + BZ_HIGH_START_OFFSET) = highmem_start;
	*(address_t *)(bzimage_setup + BZ_HIGH_SIZE_OFFSET) = highmem_size;
	if (!quiet) printf("High sections: %p bytes at %p.\n", 
			   highmem_size, highmem_start);
    } else
	highmem_start = HIGHMEM_START;

    /* Write out header and trampoline */
    if (fseek(fp, 0, SEEK_SET) < 0) {
        printf("Fatal: error seeking in output file: %s\n", 
               strerror(errno));
        exit(1);
    }
    if (fwrite(bzimage_bootsect, sizeof(bzimage_bootsect), 1, fp) != 1) {
        printf("Fatal: error writing to output file: %s\n", 
               strerror(errno));
        exit(1);
    }
    if (fwrite(bzimage_setup, sizeof(bzimage_setup), 1, fp) != 1) {
        printf("Fatal: error writing to output file: %s\n", 
               strerror(errno));
        exit(1);
    }

    if (!quiet) printf("Wrote bzImage header: %i + %i bytes.\n", 
                       sizeof(bzimage_bootsect), sizeof(bzimage_setup));

    /* Sorted list of sections below 1MB: write them out */
    for (s = sections, i = 0; s; s = s->next) {
        if (s->start >= HIGHMEM_START) continue;
        offset = (s->start - 0x9000);
        if (fseek(fp, offset, SEEK_SET) < 0) {
            printf("Fatal: error seeking in output file: %s\n", 
                   strerror(errno));
            exit(1);
        }
        if (fwrite(s->buffer, s->size, 1, fp) != 1) {
            printf("Fatal: error writing to output file: %s\n", 
                   strerror(errno));
            exit(1);
        }
        i++;
    }

    if (!quiet) printf("Wrote %i low-memory sections.\n", i);

    /* Sorted list of sections higher than 1MB: write them out */
    for (s = sections, i = 0; s; s = s->next) {
        if (s->start < HIGHMEM_START) continue;
        offset = (s->start - highmem_start) + BZ_SETUP_OFFSET;
        if (fseek(fp, offset, SEEK_SET) < 0) {
            printf("Fatal: error seeking in output file: %s\n", 
                   strerror(errno));
            exit(1);
        }
        if (fwrite(s->buffer, s->size, 1, fp) != 1) {
            printf("Fatal: error writing to output file: %s\n", 
                   strerror(errno));
            exit(1);
        }
        i++;
    }
    
    if (!quiet) printf("Wrote %i high-memory sections.\n", i);
}


/*
 *  EOF(buildimage.c)
 */
